### PR TITLE
Embedding transaction status code in receipts

### PR DESCRIPTION
**Pull request by Arachnid**
This EIP replaces the intermediate state root field of the receipt with either the contract return data and status, or a hash of that value.

### CODE REVIEW

**Comment by tjayrush** [comment]
May I ask, in option 2, why you use "`GetReturnData` and `ReturnData`, permitting nodes to fetch the status and return data" as opposed to say `GetStatus` and `GetReturnData` which would match the names of items presumably being retrieved?

**Comment by karalabe** [comment]
Given that everywhere the `0` status code means success, should we really invert it here? If you want to stick to it, then perhaps lets call it "Success" and not "Status"?

File: EIPS/eip-draft-returndata.md
Comment (karalabe) [inline EIPS/eip-draft-returndata.md]: Option 2 won't work since the data will be available only on nodes that actually ran the transaction. So you're back to the original issue you set out to solve. Also if you only have the hash of the result, there's no way to do a "success, but here's a message" scenario. Not sure if relevant.

Comment (karalabe) [inline EIPS/eip-draft-returndata.md]: I guess if you disallow "success + message", then we could use `sha(success + <empty>)` as the check in fast/light nodes to see **if** something went wrong without knowing what.

**Comment by cdetrio** [comment]
> Full nodes can provide RPCs to get a transaction return status and value by replaying the transaction, but fast nodes can only do this for nodes after their pivot point, and light nodes cannot do this at all, making a non-consensus solution impractical.

Can't light clients replay a tx by fetching the state branches (merkle proofs) for all accounts that the tx touches? Since within a block any prior tx may have modified an account, all prior tx's would also have to be replayed. But in principle, light clients are capable of replaying transactions.

**Comment by Arachnid** [comment]
@tjayrush 
> May I ask, in option 2, why you use "GetReturnData and ReturnData, permitting nodes to fetch the status and return data" as opposed to say GetStatus and GetReturnData which would match the names of items presumably being retrieved?

Typo, thanks.

@karalabe
> Given that everywhere the 0 status code means success, should we really invert it here? If you want to stick to it, then perhaps lets call it "Success" and not "Status"?

The CALL opcode returns 0 on failure and 1 on success.

File: EIPS/eip-draft-returndata.md
Comment (Arachnid) [inline EIPS/eip-draft-returndata.md]: Option 2 would work by adding a new data type for nodes to synchronise: return data. Fast synced clients would fetch this for all past receipts, and light nodes would fetch it on demand.

**Comment by cdetrio** [comment]
If something like [EIP 101](https://github.com/ethereum/EIPs/blob/master/EIPS/eip-101.md) is eventually adopted, I'm wondering how it would interact with a definition of failure as "any operation that can cause the transaction or top-level call to revert."

Under EIP 101, wouldn't the "top-level call" be the execution which pays the gas fee to the miner? Then every transaction that pays a gas fee to `block.coinbase` would be a success, and only those transactions that revert the gas payments would be failures.

**Comment by Arachnid** [comment]
After out of band discussion with Vitalik and others, I've amended this proposal to simply insert a 1 byte return status (1 for success, 0 for failure).

**Comment by axic** [comment]
I am not fully familiar with this part of the protocol, but does this EIP only refer to a P2P message or does it also influence what the `eth_transactionReceipt` RPC call do?

If the latter, wouldn't it make sense including a third option to signal revert as mentioned in https://github.com/ethereum/EIPs/pull/206#issuecomment-277825855?

File: EIPS/eip-draft-returndata.md
Comment (gumb0) [inline EIPS/eip-draft-returndata.md]: "a single byte with 0 or 1" is somewhat confusing because RLP representation of 0 (number) is empty byte array (i.e. zero bytes).
This should say either "number 0 or 1" or something like "array of size 1 with 0 or 1 byte"

cpp currently implemented it as a number, I think it's more natural.

**Comment by cdetrio** [comment]
Since this will encompass #98, how about `Replacing intermediate state roots with return data in transaction receipts` for a title?

**Comment by MicahZoltu** [comment]
The motivation section discusses including return/revert data, yet the specification section doesn't mention it.  Is this specification incomplete?  If so I recommend adding a TBD or something to make that more clear for the time being.  If not, am I missing something about how return/revert data is provided for in the transaction receipt?

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [inline EIPS/eip-draft-returndata.md]: Now the activation should be `BYZANTIUM_FORK_BLKNUM`.

**Comment by holgerd77** [comment]
Can someone add example cases for the different outcomes with concrete exemplary values to the specification?

**Comment by rjl493456442** [comment]
@Arachnid 
To sum up, from `METROPOLIS_FORK_BLKNUM ` later, the `PostState` field in receipt will been replaced by a field called `Status`, which indicates the transaction execution result.
The `Status` field is a single byte, `0x01` represents successfully execution and `0x00` represent s a failure occured(No matter accidental error or a Revert opcode called).
And the field will be added into the consensus comparison. Right?

File: EIPS/eip-draft-returndata.md
Comment (holgerd77) [reply EIPS/eip-draft-returndata.md]: @pirapira Are there actually any tests for this EIP? If so, can you mention them here? Thanks!

Comment (pirapira) [reply EIPS/eip-draft-returndata.md]: For instance, this receipt trie should be calculated using the new receipt format https://github.com/ethereum/tests/blob/develop/BlockchainTests/bcStateTests/blockhashTests.json#L22

**Comment by holgerd77** [comment]
@Arachnid This PR lacks very much a concrecte specification, are you still responsible for this and can update this since the EIP is actually getting into ``Byzantium``?

**Comment by holgerd77** [comment]
@cdetrio @axic @pirapira @vbuterin @Souptacular This PR is pretty under-specified and also still contains contradictory/unfinished information and the original creator seems to have abandoned it. Am I the only one feeling uncomfortable with this?

Go and C++ implementations have already used different names for this ([status](https://github.com/ethereum/go-ethereum/pull/15014/files#diff-cf3bf12023ca997cd6b43322a70ea69dR69) vs [statusCode](https://github.com/ethereum/cpp-ethereum/pull/4275/files#diff-82c992d147a7388a414c1154942ae0cdR568) and people are making [good guesses](https://github.com/ethereum/EIPs/pull/658#issuecomment-323258614) what the final specification actually is.

Can someone of you guys please take on this (optimally BEFORE the ``Byzantium`` HF? :-))?

Specifically I see the following points:

- What is the name of the status field?
- Update contradictory/outdated info (e.g. in the ``Abstract`` section)
- Is the old intermediate state root field kept (still not sure about this one)
- At least one concrete example with real values
- Optimally references to test cases covering this

I know this is not the most complex EIP. A "the experts will know how to interpret this" attitude nevertheless feels a bit dangerous to me.

**Comment by Arachnid** [comment]
My understanding from All Core Devs was that 98 was being amended to include the functionality from this EIP, and that this one can be discarded.

The name of the status field is fairly irrelevant - implementations can call it whatever they like.

**Comment by holgerd77** [comment]
So this would be https://github.com/ethereum/EIPs/issues/98?

Reading from a [comment](https://github.com/ethereum/EIPs/issues/98#issuecomment-318670322) from @Souptacular, this seems to be the other way around?

**Comment by karalabe** [comment]
Apparently #98 was modified to include this EIP in it. Originally my understanding was that 98 is discarded in favor o this. Either way, this and Option 3 in 98 seems to be the same thing.

**Comment by karalabe** [comment]
TL;DR Drop intermediate root, add status byte in place.

**Comment by pirapira** [comment]
@karalabe @Arachnid [EIPs/README](https://github.com/ethereum/EIPs#accepted-eips-planned-for-adoption) still points to this PR as an accepted EIP.  The link needs to be replaced if this PR is obsolete.

**Comment by karalabe** [comment]
AFAIK this PR is valid.

**Comment by Souptacular** [comment]
@holgerd77 said:

> Reading from [a comment](https://github.com/ethereum/EIPs/issues/98#issuecomment-318670322) from @Souptacular, this seems to be the other way around?

My comment is correct afaik, but the language I used could have been better. The contents of EIP 98 ([at this link](https://github.com/ethereum/EIPs/issues/98)) have been incorporated into this pull request [658](https://github.com/ethereum/EIPs/pull/658). However, the name EIP 98 is still the name of the EIP. We were using a notation EIP 98/658 for some time until we decided on an EIP number to officially represent this EIP and we chose 98 because that is what this has historically been called. So although the pull request number in GitHub is 658 this EIP written in this PR is EIP 98. There is not an "obsolete" EIP number, just a decision to use EIP 98 to represent the EIP being written at URL 658. Further, GitHub issue number 98 should no longer be changed or commented on since all discussion of EIP 98 should occur in PR 658 (the thread we are in now).

**Comment by arkpar** [comment]
I suggest naming this EIP-658 and leaving EIP-98 as it is. The two EIPs are quite different. 658 is not a continuation of 98, but more of an alternative. EIP-98 has been implemented in Parity and enabled on the kovan testnet a while ago.

**Comment by Arachnid** [comment]
I've updated the EIP to fix the obsolete text in the abstract and motivation sections.

File: EIPS/eip-draft-returndata.md
Comment (veox) [inline EIPS/eip-draft-returndata.md]: typo: iff

Comment (gumb0) [reply EIPS/eip-draft-returndata.md]: I believe it's intended ("if and only if")

File: EIPS/eip-draft-returndata.md
Comment (veox) [inline EIPS/eip-draft-returndata.md]: As far as I understand (from https://github.com/ethereum/EIPs/pull/658#issuecomment-326594093), this particular EIP will be given number 98. If so, it will be confusing if this EIP references EIP 98 as if it was a different one.

File: EIPS/eip-draft-returndata.md
Comment (veox) [reply EIPS/eip-draft-returndata.md]: [Here's a transaction](https://etherscan.io/tx/0x842e89d44f1de6cd26bca31d492932bfb4222261c08918bd7fcc39074b22fc2d) that consumed all gas and didn't fail.

It was never possible to make the assumption to begin with. ;)

Comment (Arachnid) [reply EIPS/eip-draft-returndata.md]: It wasn't, but people still did.

**Comment by LefterisJP** [comment]
Reading this EIP, I am still not sure how in a post-Byzantium world we can query this return status? We are using geth 1.7 and ropsten with the Byzantium fork activated and don't get any new field in the response of `eth_getTransactionReceipt`.

Any ideas?

**Comment by rjl493456442** [comment]
Because the status field hasn't been added to rpc response yet. I have purposed an [EIP](https://github.com/ethereum/EIPs/pull/699) to add the `statusCode` to response. But no feedback received.

**Comment by MicahZoltu** [comment]
What is the status of this?  Some sources imply that this will make it into Byzantium, yet the PR does not include the "implementations" section and Byzantium is about a week off.

**Comment by Souptacular** [comment]
@MicahZoltu I don't believe this is included in Byzantium. I will double check.

**Comment by MicahZoltu** [comment]
Ah, I think I mistook https://github.com/ethereum/EIPs/pull/211 for this one.  Sorry!

**Comment by Arachnid** [comment]
I've updated the text, removing 'a single byte', to reflect it as implemented for Byzantium.

**Comment by LefterisJP** [comment]
> I don't believe this is included in Byzantium. I will double check.

It is included.

File: EIPS/eip-draft-returndata.md
Comment (veox) [reply EIPS/eip-draft-returndata.md]: Should read fine IMO if `, already obsoleted by EIP98,` is removed.

IMO the following ` (1 for success, 0 for failure)` can be dropped, too, since this detail is elaborated on in section Specification, just a few lines below.

**Comment by MicahZoltu** [comment]
It appears that at some point this EIP lost the "return the return data of the contract execution" part along the way and turned into just "return 0 or 1".  Did returning transaction results in the receipt make it into another EIP or did it just get dropped on the floor?  Having access to return values from transactions was what I was really looking forward to in this PR, so I'm sad that it didn't make it.  :'(

**Comment by LefterisJP** [comment]
@MicahZoltu +1.

 I wonder how will the end-users be able to access the revert data at the end of day? Solidity already [has an issue](https://github.com/ethereum/solidity/issues/1686) to allow the user to specify them when throwing an exception.

**Comment by Arachnid** [comment]
@MicahZoltu @LefterisJP After consultation with others, I dropped it because of concerns about DoS and spam opportunities; return data isn't charged for (except for memory expansion) but if it's part of consensus, would have to be stored indefinitely with receipts. It's still a possibility to add this to later forks, but I didn't want to rush in something that wasn't fully thought out.

Clients can still provide return data by re-executing the transaction, of course.

**Comment by MicahZoltu** [comment]
> Clients can still provide return data by re-executing the transaction, of course.

Correct me if I'm wrong, but only full nodes can provide return-data by re-execution right?  In order for the return data to be available to light clients it must be part of the transaction receipt.

I understand your desire for cutting down this EIP to the bare minimum, I'm generally a fan of small targeted EIPs.  Unfortunately, I am not familiar enough with the internals of the Ethereum protocol to draft a new EIP for getting return data added to receipts.  Is anyone who is interested in drafting such a PR?  dApp authoring is made notably more difficult without return data being available in the receipt, and it means that using light clients for dApps is often not possible since light clients can't re-run transactions locally (as I understand it).

**Comment by Arachnid** [comment]
> Correct me if I'm wrong, but only full nodes can provide return-data by re-execution right? In order for the return data to be available to light clients it must be part of the transaction receipt.

For a light client to get it without it being part of consensus, it'd either need to trust the light server, or be supplied with proofs of all the data that went into executing the transaction in question.

I'm not sure whether including return data in receipts or providing proof-of-execution is the better approach; I can see arguments in favor of both.

> I understand your desire for cutting down this EIP to the bare minimum, I'm generally a fan of small targeted EIPs. Unfortunately, I am not familiar enough with the internals of the Ethereum protocol to draft a new EIP for getting return data added to receipts. Is anyone who is interested in drafting such a PR? dApp authoring is made notably more difficult without return data being available in the receipt, and it means that using light clients for dApps is often not possible since light clients can't re-run transactions locally (as I understand it).

A copy of this EIP with different requirements would basically be the bare minimum. To address the DoS issue, it would likely also need to specify some form of gas cost for return data that ends up in receipts.

Now would be a good time to start thinking about this, since we've got plenty of time before the next fork.

**Comment by pipermerriam** [comment]
@Arachnid given that Byzantium is live can this be merged?

File: EIPS/eip-draft-returndata.md
Comment (jamesray1) [reply EIPS/eip-draft-returndata.md]: Remind me, what's RPCs?

**Comment by Arachnid** [comment]
@pipermerriam As soon as someone approves it. :)

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [inline EIPS/eip-draft-returndata.md]: Status should now be `Accepted`.

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [inline EIPS/eip-draft-returndata.md]: Number 658 seems to have been assigned to this one.

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [inline EIPS/eip-draft-returndata.md]: I guess the title should now be changed to `Embedding transaction status code in receipts`.

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [reply EIPS/eip-draft-returndata.md]: Regarding the mention of `EIP98`, I would add a link to the relevant issue or pull-request. I believe the `Motivation` section can talk about the outer world rather freely.

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [inline EIPS/eip-draft-returndata.md]: I'll replace `METROPOLIS_FORK_BLKNUM` with `BYZANTIUM_FORK_BLKNUM`.

**Review by pirapira — CHANGES_REQUESTED** [review]
I commented about some formalities.  After a week of silence, I would edit the PR accordingly.

**Comment by pirapira** [comment]
I'll also change the filename.

File: EIPS/eip-draft-returndata.md
Comment (pirapira) [reply EIPS/eip-draft-returndata.md]: No, more like `Final`.

**Comment by Arachnid** [comment]
@pirapira PTAL.

**Comment by pirapira** [comment]
@Arachnid the changes look good.  Thanks.

I still need to
* [ ] change the filename
* [ ] give some time for others to respond to https://github.com/ethereum/EIPs/pull/658#discussion_r151970802

**Comment by tinybike** [comment]
Should the title of this PR be updated, to reflect that it's now an EIP for the new status field in receipts, rather than returndata?

**Comment by pirapira** [comment]
@tinybike done.

**Review by pirapira — APPROVED** [review]

**Comment by ivica7** [comment]
> After consultation with others, I dropped it because of concerns about DoS and spam opportunities; return data isn't charged for (except for memory expansion) but if it's part of consensus, would have to be stored indefinitely with receipts. It's still a possibility to add this to later forks, but I didn't want to rush in something that wasn't fully thought out.

DoS and Spam opportunities is a problem, but IMHO a good compromise would have been to allow to return an arbitrary 8-bit return code with the meaning 0=success, everything else an error. With 0/1 limitation we're only slightly better compared to earlier where we checked gasSent==gasUsed. On the UI side the message that can be presented to the user is only: "Everything is good" and "Something is wrong" -> bad user experience. I hope one day the specification for this will be extended.

**Comment by AUTHAI-bit** [comment]
60606040526000805460ff191690556004805460a060020a60ff0219169055341561002657fe5b604051606080610ec28339810160409081528151602083015191909201515b5b60048054600160a060020a03191633600160a060020a03161790555b600160a060020a03831615156100785760006000fd5b60058054600160a060020a031916600160a060020a038581169190911791829055604080516000602091820181905282517f18160ddd000000000000000000000000000000000000000000000000000000008152925194909316936318160ddd936004808501948390030190829087803b15156100f157fe5b6102c65a03f115156100ff57fe5b50506040805151600655600160a060020a038316600090815260026020522083905550600182905561013c64010000000061089761014682021704565b505b5050506101e7565b60045460009033600160a060020a039081169116146101655760006000fd5b60045474010000000000000000000000000000000000000000900460ff161561018e5760006000fd5b6004805460a060020a60ff021916740100000000000000000000000000000000000000001790556040517f6985a02210a168e66602d3235cb6db0e70f92b3ba4d376a33c0f3d9434bff62590600090a15060015b5b5b90565b610ccc806101f66000396000f300606060405236156100eb5763ffffffff60e060020a60003504166306fdde0381146100ed578063095ea7b31461017d578063158ef93e146101b057806318160ddd146101d457806323b872dd146101f65780632988e36b1461022f5780632a1eafd91461025f578063313ce567146102815780633f4ba83a146102a35780634b92738e146102c75780635c975abb1461032e57806370a08231146103525780638456cb59146103805780638da5cb5b146103a457806395d89b41146103d0578063a9059cbb14610460578063b85e0aca14610493578063dd62ed3e146104bf578063f2fde38b146104f3575bfe5b34156100f557fe5b6100fd610511565b604080516020808252835181830152835191928392908301918501908083838215610143575b80518252602083111561014357601f199092019160209182019101610123565b505050905090810190601f16801561016f5780820380516001836020036101000a031916815260200191505b509250505060405180910390f35b341561018557fe5b61019c600160a060020a0360043516602435610538565b604080519115158252519081900360200190f35b34156101b857fe5b61019c6105dd565b604080519115158252519081900360200190f35b34156101dc57fe5b6101e46105e6565b60408051918252519081900360200190f35b34156101fe57fe5b61019c600160a060020a03600435811690602435166044356105ec565b604080519115158252519081900360200190f35b341561023757fe5b61019c600160a060020a036004351661061d565b604080519115158252519081900360200190f35b341561026757fe5b6101e4610789565b60408051918252519081900360200190f35b341561028957fe5b6101e461078f565b60408051918252519081900360200190f35b34156102ab57fe5b61019c610794565b604080519115158252519081900360200190f35b34156102cf57fe5b61019c6004808035906020019082018035906020019080806020026020016040519081016040528093929190818152602001838360200280828437509496506107f295505050505050565b604080519115158252519081900360200190f35b341561033657fe5b61019c610868565b604080519115158252519081900360200190f35b341561035a57fe5b6101e4600160a060020a0360043516610878565b60408051918252519081900360200190f35b341561038857fe5b61019c610897565b604080519115158252519081900360200190f35b34156103ac57fe5b6103b4610916565b60408051600160a060020a039092168252519081900360200190f35b34156103d857fe5b6100fd610925565b604080516020808252835181830152835191928392908301918501908083838215610143575b80518252602083111561014357601f199092019160209182019101610123565b505050905090810190601f16801561016f5780820380516001836020036101000a031916815260200191505b509250505060405180910390f35b341561046857fe5b61019c600160a060020a0360043516602435610945565b604080519115158252519081900360200190f35b341561049b57fe5b6103b4610974565b60408051600160a060020a039092168252519081900360200190f35b34156104c757fe5b6101e4600160a060020a0360043581169060243516610983565b60408051918252519081900360200190f35b34156104fb57fe5b61050f600160a060020a03600435166109b0565b005b60408051808201909152600a815260b160020a692932b83aba30ba34b7b702602082015281565b600081158061056a5750600160a060020a03338116600090815260036020908152604080832093871683529290522054155b15156105765760006000fd5b600160a060020a03338116600081815260036020908152604080832094881680845294825291829020869055815186815291517f8c5be1e5ebec7d5bd14f71427d1e84f3dd0314c0f7b2291e5b200ac8c7c3b9259281900390910190a35060015b92915050565b60005460ff1681565b60015481565b60045460009060a060020a900460ff16156106075760006000fd5b6106128484846109fc565b90505b5b9392505050565b600454600090819033600160a060020a0390811691161461063e5760006000fd5b60005460ff161561064f5760006000fd5b600160a060020a03831660009081526002602052604081205411156106775760009150610781565b6005546040805160006020918201819052825160e060020a6370a08231028152600160a060020a038881166004830152935193909416936370a08231936024808301949391928390030190829087803b15156106cf57fe5b6102c65a03f115156106dd57fe5b5050604051519150508015156106f65760009150610781565b600160a060020a0383166000908152600260205260409020819055600154610724908263ffffffff610aff16565b600155604080518281529051600160a060020a038516917f8b80bd19aea7b735bc6d75db8d6adbe18b28c30d62b3555245eb67b2340caedc919081900360200190a2600154600654141561077c5761077a610b19565b505b600191505b5b5b50919050565b60065481565b601281565b60045460009033600160a060020a039081169116146107b35760006000fd5b60045460a060020a900460ff1615156107cc5760006000fd5b60005460ff1615156107de5760006000fd5b6107e6610b41565b50600190505b5b5b5b90565b600454600090819033600160a060020a039081169116146108135760006000fd5b60005460ff16156108245760006000fd5b5060005b825181101561077c57610851838281518110151561084257fe5b9060200190602002015161061d565b505b600101610828565b600191505b5b5b50919050565b60045460a060020a900460ff1681565b600160a060020a0381166000908152600260205260409020545b919050565b60045460009033600160a060020a039081169116146108b65760006000fd5b60045460a060020a900460ff16156108ce5760006000fd5b6004805460a060020a60ff02191660a060020a1790556040517f6985a02210a168e66602d3235cb6db0e70f92b3ba4d376a33c0f3d9434bff62590600090a15060015b5b5b90565b600454600160a060020a031681565b604080518082019091526003815260ec60020a6205245502602082015281565b60045460009060a060020a900460ff16156109605760006000fd5b61096a8383610bbb565b90505b5b92915050565b600554600160a060020a031681565b600160a060020a038083166000908152600360209081526040808320938516835292905220545b92915050565b60045433600160a060020a039081169116146109cc5760006000fd5b600160a060020a038116156109f75760048054600160a060020a031916600160a060020a0383161790555b5b5b50565b600160a060020a038084166000908152600360209081526040808320338516845282528083205493861683526002909152812054909190610a43908463ffffffff610aff16565b600160a060020a038086166000908152600260205260408082209390935590871681522054610a78908463ffffffff610c6916565b600160a060020a038616600090815260026020526040902055610aa1818463ffffffff610c6916565b600160a060020a03808716600081815260036020908152604080832033861684528252918290209490945580518781529051928816939192600080516020610c81833981519152929181900390910190a3600191505b509392505050565b600082820183811015610b0e57fe5b8091505b5092915050565b6000805460ff1615610b2b5760006000fd5b506000805460ff191660019081179091555b5b90565b60045460009033600160a060020a03908116911614610b605760006000fd5b60045460a060020a900460ff161515610b795760006000fd5b6004805460a060020a60ff02191690556040517f7805862f689e2f13df9f062ff482ad3ad112aca9e0847911ed832e158c525b3390600090a15060015b5b5b90565b600160a060020a033316600090815260026020526040812054610be4908363ffffffff610c6916565b600160a060020a033381166000908152600260205260408082209390935590851681522054610c19908363ffffffff610aff16565b600160a060020a03808516600081815260026020908152604091829020949094558051868152905191933390931692600080516020610c8183398151915292918290030190a35060015b92915050565b600082821115610c7557fe5b508082035b929150505600ddf252ad1be2c89b69c2b068fc378daa952ba7f163c4a11628f55a4df523b3efa165627a7a723058206e98be0ec21aa3ed0779e909a929dbf521053a9c0f6f4fc25da39419d4a747ca002900000000000000000000000048c80f1f4d53d5951e5d5438b54cba84f29f32a50000000000000000000000000000000000000000000000000de0b6b3a7640000000000000000000000000000e1e212c353f7a682693c198ba5ff85849f8300cc

**Comment by AUTHAI-bit** [comment]
What out Data?

**Comment by AUTHAI-bit** [comment]
Gat all u

File: EIPS/eip-draft-returndata.md
Comment (Narcosbitcoin) [inline EIPS/eip-draft-returndata.md]: Duplicate of #

**Review by ghost — APPROVED** [review]

**Review by phuttipon — COMMENTED** [review]
Duplicate of #

**Review by jokerksa12 — COMMENTED** [review]
commented on this pull request.

**Comment by joe192** [comment]
diff --git a/EIPS/eip-658.md b/EIPS/eip-658.md
new file mode 100644
index 00000000000..09afc15ec28
--- /dev/null
+++ b/EIPS/eip-658.md
@@ -0,0 +1,31 @@
+## Preamble
+
+    EIP: 658
+    Title: Embedding transaction status code in receipts
+    Author: Nick Johnson <nick@ethereum.org>
+    Type: Standard Track
+    Category Core
+    Status: Final
+    Created: 2017-06-30
+    Requires: 140
+    Replaces: 98
+
+
+## Abstract
+This EIP replaces the intermediate state root field of the receipt with a status code indicating if the top-level call succeeded or failed.
+
+## Motivation
+With the introduction of the REVERT opcode in EIP140, it is no longer possible for users to assume that a transaction failed iff it consumed all gas. As a result, there is no clear mechanism for callers to determine whether a transaction succeeded and the state changes contained in it were applied.
+
+Full nodes can provide RPCs to get a transaction return status and value by replaying the transaction, but fast nodes can only do this for nodes after their pivot point, and light nodes cannot do this at all, making a non-consensus solution impractical.
+
+Instead, we propose to replace the intermediate state root, already obsoleted by EIP98, with the return status (1 for success, 0 for failure). This both allows callers to determine success status, and remedies the previous omission of return data from the receipt.
+
+## Specification
+For blocks where block.number >= BYZANTIUM_FORK_BLKNUM, the intermediate state root is replaced by a status code, 0 indicating failure (due to any operation that can cause the transaction or top-level call to revert) and 1 indicating success.
+
+## Rationale
+This constitutes a minimal possible change that permits fetching the success/failure state of transactions, preserving existing capabilities with minimum disruption or additional work for Metropolis.
+
+## Copyright
+Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).
@Joel192

**Review by Simpdaddyjosh — APPROVED** [review]

**Comment by Cplus360** [comment]
> > After consultation with others, I dropped it because of concerns about DoS and spam opportunities; return data isn't charged for (except for memory expansion) but if it's part of consensus, would have to be stored indefinitely with receipts. It's still a possibility to add this to later forks, but I didn't want to rush in something that wasn't fully thought out.
> 
> 
> 
> @Arachnid DoS and Spam opportunities are a problem, but IMHO a good compromise would have been to allow to return an arbitrary 8-bit return code with the meaning 0=success, everything else an error. With 0/1 limitation we're only slightly better compared to earlier where we checked gasSent==gasUsed. On the UI side the message that can be presented to the user is only: "Everything is good" or "Something is wrong" -> bad user experience. I hope one day the specification for this will be extended.

Great news

**Review by Simpdaddyjosh — APPROVED** [review]

**Comment by Cplus360** [comment]
Other options?

**Comment by DCMTOKEN** [comment]
> Other options?

I prefer to use “Success”

**Comment by karalabe** [comment]
You are aware that you are discussing on a 6 year old PR that cannot change, and the decision is embedded in the Ethereum protocol, so the only way to change it is via an EIP?

**Comment by Mizanur1988** [comment]
> This EIP replaces the intermediate state root field of the receipt with either the contract return data and status, or a hash of that value.
0x19dd7d13c1f34078e92db1c16d01b55c1930d11222adc32db8359994b8891a7b

**Review by Simpdaddyjosh — APPROVED** [review]
Am I doing this right? Need some guidance

**Comment by mozz30-tech** [comment]
> This EIP replaces the intermediate state root field of the receipt with either the contract return data and status, or a hash of that value.

**Review by mozz30-tech — APPROVED** [review]

**Review by dropkick1325 — APPROVED** [review]

**Review by aramirez510 — COMMENTED** [review]
Ok

**Comment by Hliwa1990** [comment]
> Ok

**Review by Qawtr — COMMENTED** [review]
EIPS/eip-658.md

**Comment by verified-ab** [comment]
0xfcea819e17f816f0ec48592466d26a007977e0ab396030e77df37a7b8655a59f

**Comment by Musthaq22** [comment]
Arachnid:returndata

**Review by Musthaq22 — COMMENTED** [review]
EIPS/eip-658.md

**Pull request merged**
